### PR TITLE
Initial support for memory operations in reverse mode

### LINKED ISSUE #654: Compiler error related to `new` and/or pointers
**Minimal example**: Compile the following code with Clang++ with the clad plugin:
```c++
#include <iostream>
#include "clad/Differentiator/Differentiator.h"

struct T {};

double f(double x){
  T* y = new T{};
  return x;
}

int main(){
  auto f_dx = clad::gradient(f, "x");
  double d;
  f_dx.execute(3,&d);
}
```
**Expected behavior**: Code compiles.

**Observed behavior**: Compiler error:
```
heapallocation.cpp:7:10: warning: attempted to differentiate unsupported statement, no changes applied
  T* y = new T{};
         ^
clang: /home/aehle/clad/src/lib/Differentiator/StmtClone.cpp:500: clang::Stmt* clad::utils::StmtClone::VisitStmt(clang::Stmt*): Assertion `0 && "clone not fully implemented"' failed.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace, preprocessed source, and associated run script.
Stack dump:
...
```

**Further information**: An analogous error message shows up if the type `T` is replaced by e.g. `char` or `double`. 

**Setup**: Clad built from the GitHub master branch at commit https://github.com/vgvassilev/clad/commit/1bdd261ea71163b3e8074dca4018e94a481a1c50 on Ubuntu 22.04 with Clang 11.

### PR DESCRIPTION
**Pull request by vaithak**
This does not add complete support for `new` and `delete` operators, there are many examples that can be constructed which will fail or crash. However, it covers the mostly-used case of allocating memory in the beginning and deallocating in the end.

Edit: Also, added C-style memory alloc and `free` operations.

fixes #654

### CODE REVIEW

**Comment by vaithak** [comment]
@parth-07 Agreed, this is in no way a perfect solution, but meant as a fix for preventing clad from crashing on `new` and `delete` statements. For supporting all the use cases, like the one you mentioned would require a more complex analysis which can be done incrementally when use cases/issues arrive from users.
For example, in the first case, just storing the name of the pointer variable won't be sufficient. As, we can create a new pointer variable pointing to the same location and delete using that one, instead of the one we used while calling `new` expression.

**Comment by vgvassilev** [comment]
> @parth-07 Agreed, this is in no way a perfect solution, but meant as a fix for preventing clad from crashing on `new` and `delete` statements. For supporting all the use cases, like the one you mentioned would require a more complex analysis which can be done incrementally when use cases/issues arrive from users. For example, in the first case, just storing the name of the pointer variable won't be sufficient. As, we can create a new pointer variable pointing to the same location and delete using that one, instead of the one we used while calling `new` expression.

Is there a way to diagnose the unsupported cases?

**Comment by parth-07** [comment]
> For supporting all the use cases, like the one you mentioned would require a more complex analysis which can be done incrementally when use cases/issues arrive from users.

Iteratively adding features when they are required is of course the preferred way. However, we should have an idea on how to support the missing features, otherwise, later on, we may need to change the initial groundwork.

> For example, in the first case, just storing the name of the pointer variable won't be sufficient. As, we can create a new pointer variable pointing to the same location and delete using that one, instead of the one we used while calling new expression.

Yes, the first case is easier to resolve. But the second case, `delete` calls across function boundaries, can be troublesome. 

> but meant as a fix for preventing clad from crashing on new and delete statements.

If the goal for now is simply to prevent clad from crashing. Then perhaps we can simply ignore the `delete` statements. I don't think doing manual memory allocation in a mathematical function is a common case.

File: lib/Differentiator/ReverseModeVisitor.cpp [hunk 1]
```diff
@@ -2573,6 +2595,11 @@ Expr* getArraySizeExpr(const ArrayType* AT, ASTContext& context,
     bool isDerivativeOfRefType = VD->getType()->isReferenceType();
     VarDecl* VDDerived = nullptr;
     bool isPointerType = VD->getType()->isPointerType();
+    bool isInitializedByNewExpr = false;
+    // Check if the variable is pointer type and initialized by new expression
+    if (isPointerType && (VD->getInit() != nullptr) &&
```
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2600]: clang-tidy suggested to have explicit `!= nullptr` in one of the resolved comments above.

**Comment by vaithak** [comment]
> If the goal for now is simply to prevent clad from crashing. Then perhaps we can simply ignore the `delete` statements. I don't think doing manual memory allocation in a mathematical function is a common case.

I think it is better to support it minimally than to not support it at all. I don't see a way of how the analysis of such complex cases will look (even for cases like for loops or if conditions), but even if we have to remove the work in this commit entirely for a better solution, I don't see that as an issue.

File: include/clad/Differentiator/BuiltinDerivatives.h [hunk 2]
```diff
@@ -196,6 +196,26 @@ CUDA_HOST_DEVICE void clamp_pullback(const T& v, const T& lo, const T& hi,
 #endif
 
 } // namespace std
+
+ValueAndPushforward<void*, void*> malloc_pushforward(size_t sz, size_t d_sz) {
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply include/clad/Differentiator/BuiltinDerivatives.h]: We need to mark this function `inline`?

[the remark this answers is not in the document]
Comment (vaithak) [reply include/clad/Differentiator/BuiltinDerivatives.h]: yup, clang-tidy seems to be suggesting so.

File: lib/Differentiator/CladUtils.cpp [hunk 3]
```diff
@@ -641,5 +641,19 @@ namespace clad {
              isa<ObjCBoolLiteralExpr>(E) || isa<CXXBoolLiteralExpr>(E) ||
              isa<GNUNullExpr>(E);
     }
+
+    bool IsMemoryAllocationFunction(const clang::FunctionDecl* FD) {
```
Comment (vgvassilev) [inline lib/Differentiator/CladUtils.cpp:645]: Can we use `FunctionDecl::getMemoryFunctionKind()`?

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/ReverseModeVisitor.cpp:2642]: We should fix clang-tidy if possible...

File: lib/Differentiator/CladUtils.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/CladUtils.cpp:645]: In its source code, I don't see `malloc` being checked: https://clang.llvm.org/doxygen/Decl_8cpp_source.html#l04356.

Comment (vgvassilev) [reply lib/Differentiator/CladUtils.cpp:645]: I see `Builtin::BImalloc`...

Comment (vaithak) [reply lib/Differentiator/CladUtils.cpp:645]: Tried this, it didn't work.

Comment (vgvassilev) [reply lib/Differentiator/CladUtils.cpp:645]: It's probably worth trying to understand why as that'd be a design choice that will be extended in future without a lot of thinking ;)

Comment (vaithak) [reply lib/Differentiator/CladUtils.cpp:645]: Can you point the line number of link to the line where you found `BImalloc`, I still can't find it in the condition.

**Comment by parth-07** [comment]
> I think it is better to support it minimally than to not support it at all. 

We will be able to support the differentiation of functions that use `new` and `delete` if we ignore the `delete` statements. By ignoring `delete` statement, I mean that no statement will be generated in the adjoint code for a `delete` statement in the primal code. Differentiation will produce correct results even if there is a double `delete` or if there's `delete` across function boundaries. We basically have to choose between:

- Ignore `delete` statements and have correct derivatives at the cost of memory leaks in the generated code.
- Generate `delete` statements in the generated derivative code to avoid memory leaks. In this case, only the basic `delete` statement usage will work correctly. Any advance usage may result in Clad crashing, or an undefined behavior.

> I don't see a way of how the analysis of such complex cases will look (even for cases like for loops or if conditions), but even if we have to remove the work in this commit entirely for a better solution in future, I don't see that as an issue.

Yes, it's not a big issue. However, I believe more effort will be required in this way. I was just giving my two cents. Please continue with the approach that you find best.

**Comment by vaithak** [comment]
> We will be able to support the differentiation of functions that use `new` and `delete` if we ignore the `delete` statements. By ignoring `delete` statement, I mean that no statement will be generated in the adjoint code for a `delete` statement in the primal code. Differentiation will produce correct results even if there is a double `delete` or if there's `delete` across function boundaries. We basically have to choose between:
> 
> * Ignore `delete` statements and have correct derivatives at the cost of memory leaks in the generated code.
> * Generate `delete` statements in the generated derivative code to avoid memory leaks. In this case, only the basic `delete` statement usage will work correctly. Any advance usage may result in Clad crashing, or an undefined behavior.
> 

Ah, I misunderstood the ignoring delete statements part, thanks for clarifying. I guess we should let the user have some options to choose from here, some users prefer one over the other.
So, how about we provide a config option to the user? Something like, `-XClang -enable-experimental-dealloc` ?

File: lib/Differentiator/CladUtils.cpp (hunk 3 above)
Comment (vaithak) [reply lib/Differentiator/CladUtils.cpp:645]: Changed this to use the `getBuiltinID()` method and compare with defined macros like BImalloc, ...

File: lib/Differentiator/CladUtils.cpp [hunk 4]
```diff
@@ -641,5 +641,19 @@
...
+        return true;
+      if (FD->getBuiltinID() == Builtin::BIcalloc)
+        return true;
+      if (FD->getBuiltinID() == Builtin::BIrealloc)
+        return true;
+      return false;
+    }
+
+    bool IsMemoryDeallocationFunction(const clang::FunctionDecl* FD) {
+      return FD->getBuiltinID() == Builtin::BIfree;
```
[the remark this answers is not in the document]
Comment (vgvassilev) [reply lib/Differentiator/CladUtils.cpp:656]: Looks like this was introduced in llvm/llvm-project@7119f76c47797391fddcfcb60f29e70a2b424713 which is available on clang 15 onward...

**Comment by vgvassilev** [comment]
> > We will be able to support the differentiation of functions that use `new` and `delete` if we ignore the `delete` statements. By ignoring `delete` statement, I mean that no statement will be generated in the adjoint code for a `delete` statement in the primal code. Differentiation will produce correct results even if there is a double `delete` or if there's `delete` across function boundaries. We basically have to choose between:
> > 
> > * Ignore `delete` statements and have correct derivatives at the cost of memory leaks in the generated code.
> > * Generate `delete` statements in the generated derivative code to avoid memory leaks. In this case, only the basic `delete` statement usage will work correctly. Any advance usage may result in Clad crashing, or an undefined behavior.
> 
> Ah, I misunderstood the ignoring delete statements part, thanks for clarifying. I guess we should let the user have some options to choose from here, some users would prefer one over the other depending on their applications. So, how about we provide a config option to the user? Something like, `-XClang -enable-experimental-dealloc` ?

I was wondering if we could emit the delete statements when the allocation is in the function we differentiate and produce a warning if the allocation/deallocation is across function boundaries?

**Comment by vaithak** [comment]
> I was wondering if we could emit the delete statements when the allocation is in the function we differentiate and produce a warning if the allocation/deallocation is across function boundaries?

But, there are many cases where even if the delete(s) are in the same function, the current implementation will output wrong statements, for ex. - multiple new and delete combinations (as exemplified by Parth), delete(s) inside for/if conditions and maybe many more.

File: lib/Differentiator/ReverseModeVisitor.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (vaithak) [reply lib/Differentiator/ReverseModeVisitor.cpp:2652]: fixed this 👍🏼

**Comment by vaithak** [comment]
Turns out there was some issue when zero initializing a newly created array using `new double[2]()`, it took a good debugging session to figure out the root cause, but this is fixed now.

**Comment by vaithak** [comment]
All tests are passing finally 🎉

File: test/ForwardMode/Pointer.C [hunk 5]
```diff
@@ -110,16 +110,84 @@ double fn5(double i, double j) {
...
+// CHECK-NEXT:     *_d_p = _d_i;
+// CHECK-NEXT:     *p = i;
+// CHECK-NEXT:     {{(clad::)?}}ValueAndPushforward<void *, void *> _t1 = clad::custom_derivatives::calloc_pushforward(1, sizeof(T), 0, sizeof(T));
+// CHECK-NEXT:     T *_d_t = (T *)_t1.pushforward;
+// CHECK-NEXT:     T *t = (T *)_t1.value;
+// CHECK-NEXT:     _d_t->i = _d_i;
+// CHECK-NEXT:     t->i = i;
+// CHECK-NEXT:     double _d_res = *_d_p + _d_t->i;
+// CHECK-NEXT:     double res = *p + t->i;
+// CHECK-NEXT:     unsigned {{(int|long)}} _t2 = sizeof(double);
```
Comment (vgvassilev) [inline test/ForwardMode/Pointer.C:164]: I was wondering why we are not writing out `size_t` here?

Comment (vaithak) [reply test/ForwardMode/Pointer.C:164]: not sure about this

**Comment by vgvassilev** [comment]
@parth-07 and I chatted a little and we'd like to move forward. Once we get some more experience then we should decide if we want to relax the delete operations in favor of more memory leaks.

**Pull request merged**